### PR TITLE
Minor EE Cache changes

### PR DESCRIPTION
**Pull request by prafullpcsx2**
This seems to help games that require EE Cache. Please test **[Ice Age 2](http://wiki.pcsx2.net/index.php/Ice_Age_2:_The_Meltdown)** and **[Evolution Snowboarding](http://wiki.pcsx2.net/index.php/Evolution_Snowboarding)**. 

**Note**: [Need to enable Interpreter + EE Cache](http://ultraimg.com/images/EECache.png).

### CODE REVIEW

**Comment by refractionpcsx2** [comment]
I'll test this later :) I'm expecting nothing terrible to happen as it is restoring old code which I removed, possibly shouldn't have!

**Comment by prafullpcsx2** [comment]
@refractionpcsx2 Please do have a look at it because I haven't fully reverted your commit but slightly changed it according to your newer changes but both seemed to have same effect.

**Comment by refractionpcsx2** [comment]
Yeh that is writing back to memory, not sure if it's adding anything new to the cache, I can't remember exactly what it did >.<

**Comment by refractionpcsx2** [comment]
Do I need to do anything else for Ice Age? It still doesn't work for me

**Comment by refractionpcsx2** [comment]
ok i managed to get mine to move by forcing the cache to write back to memory on the DXLTC CACHE instruction. before this it would just freeze before the video

**Comment by prafullpcsx2** [comment]
That's good news. Yes it needed either the skip video patch to complete skip videos or oph flag hack to manually skip them. So with your new changes the game works now?

**Comment by refractionpcsx2** [comment]
Yep, the videos work and I went ingame.  If course it was slow as we expect

**Comment by prafullpcsx2** [comment]
@refractionpcsx2 Can you please update this PR with the changes you mentioned above? Also did you manage to switch to recompiler mode or got savestates working?

**Comment by refractionpcsx2** [comment]
I haven't done no, to be honest I didn't look at that.  I kinda hoped I could switch to the recompiler with the cache on but it doesn't look like it :(

I could add a function to flush the cache on switching to the recompiler. As for savestates, I didn't really want to break them again :p I guess it happens though

**Comment by refractionpcsx2** [comment]
There you go :)

**Comment by MrCK1** [comment]
I actually have Evolution Snowboarding now :) Is the difference supposed to be noticeable because I couldn't tell any difference between the Appveyor and Git Builds.

I wasn't able to switch to Recompiler as well; and the game still requires OPH/VIF1 FIFO Read hacks. Very slow obviously ;)

[Appveyor Log](http://pastebin.com/qLSZeXNQ)

File: pcsx2/Cache.cpp [hunk 1]
```diff
@@ -506,4 +536,4 @@ void CACHE() {
 }
 }		// end namespace OpcodeImpl
 
-}}
\ No newline at end of file
+}}
```
Comment (monster860) [inline pcsx2/Cache.cpp]: *screams*

**Comment by ramapcsx2** [comment]
Is this PR doing what was intended? 
@monster860 probably wanted a new line at the end of the file?

Requesting an update.

**Comment by refractionpcsx2** [comment]
I spose it's ready enough for now, it will need work later for savestate and hot swapping compatibility.

**Comment by gregory38** [comment]
could the code be replaced with some memcpy ? Or is it a perf issue ? Code is awful, I'm afraid it will break C++ aliasing rule and/or 64 bits.

**Comment by refractionpcsx2** [comment]
Maybe, but all the cache code is like this.

**Comment by MrCK1** [comment]
I don't think I really made it clear anywhere but Evolution Snowboarding doesn't need EE cache as far as I know.

The game code will tell the screen to go blank/freeze the current image at the end of a level; and doesn't resume (until you select a level through audio cues)

**Comment by refractionpcsx2** [comment]
The videos don't work without EE cache, but they do work with it on. afaik.

**Comment by prafullpcsx2** [comment]
Last time I tested Evolution Snowboarding went inagme with these changes without needing vif/oph hacks. Will test it again later.

**Comment by gregory38** [comment]
```
Maybe, but all the cache code is like this. I think refactoring is outside the scope of this PR.
```
Well, cache code reports various bad warning on 64 bits. It would be nice if code can be improved a bit, even in a separate PR.

**Comment by ramapcsx2** [comment]
May I recommend a "project" to investigate this whole cache business?
It seems odd that we can't make this work without fully emulating the CPU cache.
Also it's strictly interpreter, which limits its usefulness.

**Comment by refractionpcsx2** [comment]
Possibly, DoA2 requires cache as well (at least one version) in order to boot, not sure why.

**Comment by MrCK1** [comment]
@prafullpcsx2 You were correct about the EE cache, no hacks were necessary. This also works on latest git.

After the title screen, the 1st in-game video (a comic strip of sorts) is only played when using the EE cache.

**Comment by prafullpcsx2** [comment]
@MrCK1 : That's why I asked to check that game in first post. By the way, I think World Rally Championship 4 also requires ee cache. Copy pasting Nachbrenner's old post 

> With this game the ELF is compressed.
Load address is 00400000, code is unpacked to 00100000

loops at
EPC= 000835b8
PC/RA = 000825b8
SP=00093100
until VM crashes printing vmhack
IDA might give you a clue

TLB crashes with cpuTLBmiss 4006c4
which means whilst unpacking

> I took a closer look at the compression:
Original ELF loads at
0x00400000 - 0x0053bcac
then unpacks to
0x20100000 - 0x204a333c
I assume it crashes because RAM caching aint emulated and the memory clashes.

**Comment by sudonim1** [comment]
@refractionpcsx2 Oh about DoA2, I know why.  It has buggy DMA to its (custom) sound module on the IOP, which only works iirc because some data's actually stuck in the write-back cache and isn't included in the DMA.  I'm 99% sure happened is they forgot to flush the cache and when it didn't work, they swapped a couple of lines instead of adding a flush, it started working and they moved on.

**Comment by refractionpcsx2** [comment]
@sudonim1  Oh so it does need the cache because they buggered their DMA's up??

**Comment by sudonim1** [comment]
Yeah the patch I wrote for a couple of versions fixes this by swapping the code back around to the order it should be if they had a flush.  There's something else going on in addition in another version though.

**Comment by MrCK1** [comment]
Good news for Evolution Snowboarding. I managed to get further into the game to a point where I could save the memcard. This will allow you to get to the actual main menu of the game


If you need it for whatever, [here it is](http://www.filedropper.com/evolutionsnowboarding02)

**Comment by gregory38** [comment]
So what do we do with this PR ?

**Comment by sudonim1** [comment]
The reason it wasn't being merged was the general code quality of the cache implementation, right? Honestly this seems like a non-concern.  Even if the portability becomes a practical issue the module can be compiled out, it's not really something that affects users since it's appallingly slow on top of having to use the EE interpreter.  The whole thing probably wouldn't pass a good test suite anyway and will likely need a rewrite when we want to make it pass, at which point it can be brought up to scratch for safety and portability.

**Comment by refractionpcsx2** [comment]
Alright I'll merge this,  we can revisit the code quality later

**Pull request merged**